feat: Preserve symlinks when copying templates

AdrianFreundQC opened the pull request:
Closes #937.

This PR makes copier preserve symlinks when copying a template instead of copying the file the symlink points to.

The old behavior can be restored by setting `_preserve_symlinks` to false in `copier.yml`.

If the name of a symlink is templated it will be rendered. The targets of symlinks will only be rendered if the symlink name ends with the template suffix (`.jinja` by default).

AdrianFreundQC commented:
I have just force-pushed some fixed. I can now confirm that this works using the template in the linked issue and breaks no existing tests.

I haven't yet written automated tests for this.

yajo commented:
I think we'll break a lot of workflows if this is not optional (even if it's enabled by default, I think there should be a way to disable it).

@sisp @pawamoy what do you think?

pawamoy commented:
I don't know about a lot of workflows, but just recently we had someone with a use-case for symlinks being rendered.

AdrianFreundQC commented:
Pushed again. Tests should no longer fail on python <3.10.

I can try making this an option. Probably only next week though.

sisp commented:
> This is actually a feature that allows you to reuse files in some ways and reduce code duplication. However indeed it would make sense to make it optional.
>
> &mdash; https://github.com/copier-org/copier/issues/937#issuecomment-1405775180

TBH, I would have expected that symlinks remain symlinks after generating a project. Would it be possible to "reuse files [...] and reduce code duplication" via [Jinja "include"](https://jinja.palletsprojects.com/en/3.1.x/templates/#include) instead? If it works, it only works for text files though but not for binary files. Or [Copier "tasks"](https://copier.readthedocs.io/en/latest/configuring/#tasks) could be used to copy files, although I admit it's less declarative than the current symlink approach or the Jinja "include" idea.

If making it optional is the way to go, I wonder whether it's sufficient to have a global setting or whether it should be configurable at a more granular level. :thinking:

sisp commented:
By the way, the current symlink behavior seems to be undocumented. This leads me to wonder:

1. How many Copier users actually rely on the current symlink behavior?
2. How bad would it be to introduce a breaking change to this undocumented feature? (Although it's certainly not a good idea to break people's templates, but relying on an undocumented feature is known to be dangerous.)

AdrianFreundQC commented:
My suggestion would be the following:
Preserve symlinks by default. This is what I would expect and what is in line with other tools such as cookiecutter. If this is too disruptive it might be possible to have it not preserve by default and change the default once 8.0 releases.

jinja templating in both the symlink name and the symlink target should be rendered (So you can for example have a symlink that points to a file with a templated name).

There should be an option in `copier.yml` to change whether or not to preserve symlinks. When it is off everything should behave as before.
I'm not sure if a CLI option should exist, but I tend towards no, as it would probably mostly lead to users breaking templates that were designed with one of the options in mind.

AdrianFreundQC commented:
I also just checked the top 20 public copier templates on github and found exactly one that relies on symlinks being replaced with the file they point to.

https://github.com/OCA/oca-addons-repo-template uses a symlink to copy the `.gitignore` from the template repo to  generated project.

Maybe a third option would be useful for some (Preserve all symlinks, preserve no symlinks, preserve only symlinks pointing to inside the generated project). Or maybe that would just be confusing and lead to more problems...

yajo commented:
> https://github.com/OCA/oca-addons-repo-template uses a symlink to copy the `.gitignore` from the template repo to generated project.

I happen to maintain that template, and some others with at least that same purpose. However that can be easily changed to using import.

I'd just add an option to copier.yaml: `_symlinks`  with choices `as_symlinks` or `as_files`.



> jinja templating in both the symlink name and the symlink target should be rendered (So you can for example have a symlink that points to a file with a templated name).

Makes total sense.

yajo commented on copier/main.py:
```diff
@@ -513,6 +517,34 @@ def _render_file(self, src_abspath: Path) -> None:
             dst_abspath.chmod(src_mode)
 
+    def _render_symlink(self, src_abspath: Path) -> None:
+        assert src_abspath.is_absolute()
+        src_relpath = src_abspath.relative_to(self.template_copy_root)
+        dst_relpath = self._render_path(src_relpath)
+        if dst_relpath is None:
+            return
+        dst_abspath = Path(self.subproject.local_abspath, dst_relpath)
+        if dst_abspath.is_dir():
+            return
+        src_mode = src_abspath.lstat().st_mode
+
+        src_target = readlink(src_abspath)
+        dst_target = self._render_path(src_target)
+        if dst_target is None:
+            return
```
imho the symlink should be copied raw always, when `preserve_symlinks == True`.

If it is absolute, it should be copied absolute. If it is relative, then relative. Template maintainers can create the symlink as they want, so it's better to lend them that responsability. Just document it and test it.

yajo commented on copier/template.py:
```diff
@@ -435,6 +435,15 @@ def templates_suffix(self) -> str:
             return DEFAULT_TEMPLATES_SUFFIX
         return result
 
+    @cached_property
+    def preserve_symlinks(self) -> bool:
+        """Get whether to preserve symlinks
+
+        By default: `True`
+
+        See [preserve_symlinks][]"""
+        return bool(self.config_data.get("preserve_symlinks", True))
```
> I'd just add an option to copier.yaml: `_symlinks` with choices `as_symlinks` or `as_files`.

Sorry, I didn't see this code. Your option name seems better to me. And I like it more that it is a boolean. Please leave it as you've done, thanks!

yajo commented on copier/tools.py:
```diff
@@ -178,3 +178,17 @@ def cleanup(self):
+
+
+def readlink(link: Path) -> Path:
+    """A custom version of os.readline/pathlib.Path.readline.
+
+    pathlib.Path.readline is what we ideally would want to use, but it is only available on python>=3.9.
+    os.readline doesn't support Path and bytes on Windows for python<3.8
```
```suggestion
    """A custom version of os.readline/pathlib.Path.readlink.

    pathlib.Path.readlink is what we ideally would want to use, but it is only available on python>=3.9.
    os.readlink doesn't support Path and bytes on Windows for python<3.8
```

AdrianFreundQC commented:
I think this PR should be ready now.

yajo commented on copier/main.py:
```diff
@@ -320,7 +321,10 @@ def _render_allowed(
         try:
-            previous_content = dst_abspath.read_bytes()
+            if is_symlink:
+                previous_content = bytes(readlink(dst_abspath))
```
Seems less error-prone:

```suggestion
                previous_content = readlink(dst_abspath)
```

yajo commented on copier/main.py:
```diff
@@ -300,6 +300,7 @@ def _render_allowed(
         dst_relpath: Path,
         is_dir: bool = False,
+        is_symlink: bool = False,
         expected_contents: bytes = b"",
```
```suggestion
        expected_contents: Union[bytes, str] = b"",
```

AdrianFreundQC replied:
I used `Union[bytes, Path]` instead.

yajo commented on copier/main.py:
```diff
@@ -513,6 +517,36 @@ def _render_file(self, src_abspath: Path) -> None:
+
+        if not self._render_allowed(
+            dst_relpath,
+            expected_contents=bytes(dst_target),
```
```suggestion
            expected_contents=dst_target,
```

yajo commented on copier/main.py:
```diff
@@ -513,6 +517,36 @@ def _render_file(self, src_abspath: Path) -> None:
 
+    def _render_symlink(self, src_abspath: Path) -> None:
+        assert src_abspath.is_absolute()
+        src_relpath = src_abspath.relative_to(self.template_copy_root)
```
I think I commented it previously... Absolute links should be kept absolute IMHO.

AdrianFreundQC replied:
This doesn't change the target of the symlink. This is the path where the symlink itself is.
The handling of the symlink target is in the code block below

yajo commented on copier/main.py:
```diff
@@ -513,6 +517,36 @@ def _render_file(self, src_abspath: Path) -> None:
             dst_abspath.write_bytes(new_content)
             dst_abspath.chmod(src_mode)
 
+    def _render_symlink(self, src_abspath: Path) -> None:
```
Please add a docstring to the method, explaining clearly what it does. It will help my future self when maintaining this 😄

pawamoy commented on copier/main.py:
```diff
@@ -513,6 +517,36 @@ def _render_file(self, src_abspath: Path) -> None:
+
+        src_target = readlink(src_abspath)
+        if src_abspath.name.endswith(self.template.templates_suffix):
+            dst_target = Path(self._render_string(str(src_target)))
```
Yep, docs absolutely needed, especially since there's a new config option `_preserve_symlinks`.

AdrianFreundQC replied:
Is it enough to add it to the `preserve_symlinks` option documentation or should I also add it somewhere else in the docs? If so, where would it fit best?

yajo replied:
I see it OK with the docs you already added. Thanks!

yajo commented on copier/template.py:
```diff
@@ -435,6 +435,15 @@ def templates_suffix(self) -> str:
 
+    @cached_property
+    def preserve_symlinks(self) -> bool:
+        """Get whether to preserve symlinks
+
+        By default: `True`
+
+        See [preserve_symlinks][]"""
```
To fit with style:

```suggestion
        """Know if Copier should preserve symlinks when rendering the template.

        See [preserve_symlinks][].
    """
```

yajo commented on copier/template.py:
```diff
@@ -435,6 +435,15 @@ def templates_suffix(self) -> str:
+        By default: `True`
+
+        See [preserve_symlinks][]"""
+        return bool(self.config_data.get("preserve_symlinks", True))
```
I think that we should default to `False`.

There are a lot of templates that use this feature as it was implemented before. At least I have a few myself. I also think it's a good default.

```suggestion
        return bool(self.config_data.get("preserve_symlinks", False))
```

This will affect some of the tests.

AdrianFreundQC replied:
Changed default to false and adjusted tests. Most tests test the default (False) behavior. All tests testing the True behavior are in `test_symlinks.py`

yajo commented on docs/configuring.md:
```diff
@@ -854,6 +854,18 @@ Run but do not make any changes.
+
+-   Format: `bool`
+-   CLI flags: N/A
+-   Default value: `True`
```
As above:

```suggestion
-   Default value: `False`
```

yajo reviewed (changes requested):
Thanks for this enhancement. It is very well structured! And special thanks for the tests!

Some comments below.

AdrianFreundQC commented:
I addressed all the review comments and rebased because of a conflict. I'm not yet sure why one test fails when run with `flake check`, but passes when run  with `poe test`, but I can reproduce it locally and am looking into it.

AdrianFreundQC commented:
Tests pass again. nix flake didn't preserve the symlink in the demo template while copying it to the nix environment. The test couldn't preserve the symlink because it wasn't there in the first place.

I removed the symlink from the demo template again and instead added more fine-grained tests, so this has 100% diff coverage again.

I'm not familiar enough with nix to know if this is a bug or expected behavior.

AdrianFreundQC commented:
I just rebased this to fix the merge conflicts. Any update on when this might get merged or get another review round?

yajo reviewed (approved):
Excellent, thanks!
